handlers/wechat/wechat.go

```diff
@@ -0,0 +1,319 @@
+	"fmt"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/buger/jsonparser"
+	"github.com/garyburd/redigo/redis"
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/handlers"
+	"github.com/nyaruka/courier/utils"
+	"github.com/nyaruka/gocommon/urns"
+)
+
+var (
+	sendURL      = "https://api.weixin.qq.com/cgi-bin"
+	maxMsgLength = 1600
+	fetchTimeout = time.Second * 2
+)
+
+const (
+	configAppID     = "wechat_app_id"
+	configAppSecret = "wechat_app_secret"
+)
+
+func init() {
+	courier.RegisterHandler(newHandler())
+}
+
+type handler struct {
+	handlers.BaseHandler
+}
+
+func newHandler() courier.ChannelHandler {
+	return &handler{handlers.NewBaseHandler(courier.ChannelType("WC"), "WeChat")}
+}
+
+// Initialize is called by the engine once everything is loaded
+func (h *handler) Initialize(s courier.Server) error {
+	h.SetServer(s)
+	s.AddHandlerRoute(h, http.MethodGet, "", h.VerifyURL)
+	s.AddHandlerRoute(h, http.MethodPost, "", h.receiveMessage)
+	return nil
+}
+
+type verifyForm struct {
+	Signature string `name:"signature"`
+	Timestamp string `name:"timestamp"`
+	Nonce     string `name:"nonce"`
+	EchoStr   string `name:"echostr"`
+}
+
+// VerifyURL is our HTTP handler function for WeChat config URL verification callbacks
+func (h *handler) VerifyURL(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.Event, error) {
+	form := &verifyForm{}
+	err := handlers.DecodeAndValidateForm(form, r)
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+	}
+
+	dictOrder := []string{channel.StringConfigForKey(courier.ConfigSecret, ""), form.Timestamp, form.Nonce}
+	sort.Sort(sort.StringSlice(dictOrder))
+
+	combinedParams := strings.Join(dictOrder, "")
+
+	hash := sha1.New()
+	hash.Write([]byte(combinedParams))
+	encoded := hex.EncodeToString(hash.Sum(nil))
+
+	ResponseText := "unknown request"
+	StatusCode := 400
+
+	if encoded == form.Signature {
+		ResponseText = form.EchoStr
+		StatusCode = 200
+		go h.fetchAccessToken(channel)
+	}
+
+	w.Header().Set("Content-Type", "text/plain")
+	w.WriteHeader(StatusCode)
+	_, err = fmt.Fprint(w, ResponseText)
+	return nil, err
+}
+
+// fetchAccessToken tries to fetch a new token for our channel, setting the result in redis
+func (h *handler) fetchAccessToken(channel courier.Channel) error {
+	time.Sleep(fetchTimeout)
+
+	form := url.Values{
+		"grant_type": []string{"client_credential"},
+		"appid":      []string{channel.StringConfigForKey(configAppID, "")},
+		"secret":     []string{channel.StringConfigForKey(configAppSecret, "")},
+	}
+
+	tokenURL, _ := url.Parse(fmt.Sprintf("%s/%s", sendURL, "token"))
+	tokenURL.RawQuery = form.Encode()
+
+	req, _ := http.NewRequest(http.MethodGet, tokenURL.String(), nil)
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+	rr, err := utils.MakeHTTPRequest(req)
+	if err != nil {
+		return err
```

Comment:
We should be logging since this error isn't going anywhere otherwise. (ideally should create a channel log)